### PR TITLE
feat(ingest): powerbi # powerbi dataset schema snapshot should be ingested

### PR DESCRIPTION
**Pull request by looppi**
* Parse returned dataset columns and measures to schema fields and emit them to Datahub.
* Add switch to ingest datasets even if they're not used in that workspace.
* Allow ODBC datasets to be handled by query parsing

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Comment by looppi** [comment]
I made a decision here that datasets which are not directly used in any report or dashboard, are ingested to Datahub. In the previous implementation those datasets would have been left out. Would it be better if such action would be behind a feature flag? For our use case it was must have requirement, but I'm not sure anymore whether that is the case with general use case.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py [hunk 1]
```diff
@@ -700,8 +765,12 @@ def get_data_source(self, dataset: Dataset) -> Any:
         if self.__config.dataset_type_mapping.get(datasource.type) is not None:
             # Now set the database detail as it is relational data source
             datasource.metadata = PowerBiAPI.DataSource.MetaData(is_relational=True)
-            datasource.database = datasource_dict["connectionDetails"]["database"]
-            datasource.server = datasource_dict["connectionDetails"]["server"]
+            datasource.database = datasource_dict.get("connectionDetails", {}).get(
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi.py:768]: Thank you!

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py [hunk 2]
```diff
@@ -604,6 +609,66 @@ def get_dataset(self, workspace_id: str, dataset_id: str) -> Any:
...
+            response = requests.post(
+                dataset_query_endpoint,
+                json=payload,
+                headers={Constant.Authorization: self.get_access_token()},
+            )
+            response.raise_for_status()
+
+            results = {}
+            data = response.json()
+            rows = data["results"][0]["tables"][0]["rows"]
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi.py:644]: Can we really depend on results[0][tables][0][rows]  being there??

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py [hunk 3]
```diff
@@ -604,6 +609,66 @@ def get_dataset(self, workspace_id: str, dataset_id: str) -> Any:
...
+            for row in rows:
+                table_name = row["[Table Name]"]
+                col_name = row["[Column Name]"]
+                if table_name not in results:
+                    results[table_name] = [col_name]
+                else:
+                    results[table_name].append(col_name)
+            return results
+        except requests.exceptions.RequestException as ex:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi.py:653]: What if another exception is thrown while processing the rows? Can we except other exception types?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py [hunk 4]
```diff
@@ -141,6 +142,9 @@ class PowerBiAPIConfig(EnvBasedSourceConfigBase):
     extract_reports: bool = pydantic.Field(
         default=True, description="Whether reports should be ingested"
     )
+    extract_orphan_datasets: bool = pydantic.Field(
+        default=False, description="Whether orphan datasets should be ingested"
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi.py:146]: Minor: It might be better to describe in which cases datasets will be considered "orphaned" in the docs. I can totally see someone asking "what does orphaned mean?" while reading this doc

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py [hunk 5]
```diff
@@ -604,6 +609,66 @@ def get_dataset(self, workspace_id: str, dataset_id: str) -> Any:
...
+    def format_connection_string(data_source: Optional[Dict]) -> Optional[str]:
+        if not data_source:
+            return None
+
+        connection_string = data_source.get("connectionDetails", {}).get(
+            "connectionString", None
+        )
+        if not connection_string:
+            return None
+        return connection_string.replace("dsn=", "")
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi.py:670]: Why was this changed required?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py [hunk 6]
```diff
@@ -1029,6 +1100,16 @@ def json_to_dataset_map(scan_result: dict) -> dict:
                         f"Processing tables attribute for dataset {dataset_instance.name}({dataset_instance.id})"
                     )
 
+                    if not dataset_dict["tables"]:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi.py:1103]: So this only happens in ODBC cases? Otherwise tables is always populated?

Comment (looppi) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi.py:1103]: Currently what we've experienced, it has been that only ODBC sources are missing tables. When reflecting this against the [this PR](https://github.com/datahub-project/datahub/pull/6549) it might be that the tables are not there, but the expression is. IMO correct approach would be to use the DAX query as last resort if no other mean is available for discovering the schema.

The API does throw an error 500 sometimes, so it has not been completely reliable API.

Would it be better if this feature would be completely behind a feature flag?

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py (hunk 5 above)
Comment (looppi) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi.py:670]: This was just to make the connection string more readable and then make the dataset URNs more sensible.

File: metadata-ingestion/src/datahub/ingestion/source/powerbi.py (hunk 4 above)
Comment (looppi) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi.py:146]: You're totally correct. I'll add comment to the code and recipe at least.

**Review by mohdsiddique — CHANGES_REQUESTED** [review]

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py [hunk 7]
```diff
@@ -447,6 +449,54 @@ def get_dataset(self, workspace_id: str, dataset_id: str) -> Any:
             tables=[],
         )
 
+    def get_dataset_schema(self, dataset: PowerBIDataset) -> Dict[str, List[str]]:
```
[the remark this answers is not in the document]
Comment (looppi) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py:452]: The origin of this PR is the fact that when using ODBC datasources in PowerBI the DataHub ingestion completely skips ingestion of datasets using ODBC. This is mainly because the workspace scan result dict has `tables` as empty list. There's no column information present at the payload either.

In order to get anything in the datasets visible, I implemented the DAX query to achieve that. I suspect that using ODBC isn't that popular in the end. We're currently forced to use ODBC datasources in PowerBI.

This is `datasect_dict` contents used in https://github.com/datahub-project/datahub/pull/6440/files#diff-72724d37501e21166f538e16168db8dbec7a9b17f93e4681443335e9136e158aR920 (I redacted ID's and other sensitive data from payload, the point is that the tables list is empty).
```
{'configuredBy': 'xxxxx',
 'configuredById': 'xxxxx',
 'contentProviderType': 'PbixInImportMode',
 'createdDate': '2022-12-01T11:07:40.787',
 'datasourceUsages': [{'datasourceInstanceId': 'xxxxxxx'}],
 'description': '.... ',
 'id': 'xxxxx',
 'isEffectiveIdentityRequired': False,
 'isEffectiveIdentityRolesRequired': False,
 'name': 'xxxx',
 'tables': [],
 'targetStorageMode': 'PremiumFiles',
 'users': [.....]}

```

**Comment by jjoyce0510** [comment]
Checking in - where are we on this?

**Comment by looppi** [comment]
I slightly changed course. As I finally realized that our tenant didn't have all the metadata switches enabled, I've now implemented the schema parsing from the scan result. Initially I had good progress, but I'm still trying to fix a case where in PowerBI, a dataset has some kind of folder in it and the columns are inside the folder. I will be dropping the DAX calls and other now unnecessary stuff asap and start to work towards finishing this code.

**Comment by jjoyce0510** [comment]
Waiting on another pass from @Mohd-gslab. 

Thanks for the updates @looppi !

**Comment by looppi** [comment]
@mohdsiddique I fixed couple of issues in the code, one being Python 3.7 compatibility issue. I think this should be ready for review now. Thank you!

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/powerbi_api.py [hunk 8]
```diff
@@ -305,11 +307,42 @@ def _get_workspace_datasets(self, scan_result: Optional[dict]) -> dict:
...
                 )
 
         return dataset_map
 
+    @staticmethod
+    def get_measures(measures: List[dict]) -> List[Measure]:
+        results = []
+        for measure in measures:
+            ms = Measure(
+                name=measure.get(Constant.NAME, ""),
```
Comment (mohdsiddique) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/rest_api_wrapper/powerbi_api.py:322]: Make it optional and use None as default value. Make similar changes all other places.

**Review by mohdsiddique — CHANGES_REQUESTED** [review]

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py [hunk 9]
```diff
@@ -270,8 +295,63 @@ def to_datahub_dataset(
...
+
+            yield SchemaField(
+                fieldPath=column.name,
+                type=SchemaFieldDataType(
+                    type=powerbi_type_mapping.get(
+                        getattr(column, "data_type", ""), NullTypeClass
+                    )()
+                ),
+                nativeDataType=getattr(column, "data_type", "Unknown"),
+                description=column.description or "",
```
[the remark this answers is not in the document]
Comment (looppi) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:349]: Hi! I'm afraid we're using the measure description field currently for documentation purposes and it's crucial for us to have both column and measure descriptions visible in Datahub.

[the remark this answers is not in the document]
Comment (aezomz) [reply metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:349]: https://github.com/datahub-project/datahub/pull/7519
I have made a PR for this as well. 
I concat column.description and column.expression after taking your feedback

**Comment by hsheth2** [comment]
@looppi would it be accurate to say that https://github.com/datahub-project/datahub/pull/7519 is a superset of the changes here? I'm planning on merging that one soon cc @aezomz, but wanted to make sure that satisfies your requirements as well.

**Comment by looppi** [comment]
@hsheth2 Yes, it's correct, the #7519 includes an implementation of the PowerBI schema handling too.

I have one question though: Is there a test case in #7519 for testing the outcome of the column/measure ingestion? I couldn't find a change in the golden files related to this feature. In this PR #6440, all of the golden files changed so that the `schemaMetadata`event is visible in the golden files.

**Comment by hsheth2** [comment]
@looppi thanks for the confirmation

I'm closing this PR in favor of https://github.com/datahub-project/datahub/pull/7519
